### PR TITLE
Allow protocols to be nested in non-generic contexts

### PR DESCRIPTION
**Pull request by karwa**
I'd like to add more tests. Some guidance would be appreciated.

Does this require swift-evo? I feel that it's an "obvious" feature.

### CODE REVIEW

File: test/PrintAsObjC/protocols.swift [hunk 1]
```diff
@@ -187,6 +187,20 @@ extension NSString : A, ZZZ {}
...
 
+// CHECK-LABEL: @interface RootClass4 (SWIFT_EXTENSION(protocols))
+// CHECK-NEXT: @end
+extension RootClass4 {
+
+  // CHECK-LABEL: @protocol DefinedInExtension
+  // CHECK-NEXT: @end
+  @objc protocol DefinedInExtension {}
+
+  // CHECK-LABEL: SWIFT_PROTOCOL_NAMED("DefinedInExtension2")
```
Comment (karwa) [inline test/PrintAsObjC/protocols.swift]: I have a feeling this should be `RootClass4.DefinedInExtension2`, but I'm not certain about how the `swift_name` attribute is used. In any case the same happens for classes:

```swift
class Outer: NSObject { @objc(CustomClassName) class Inner: NSObject {} }
```
Generates the header entry:
```objc
SWIFT_CLASS_NAMED("Inner") // <-- Not "Outer.Inner"
@interface CustomClassName : NSObject
- (nonnull instancetype)init OBJC_DESIGNATED_INITIALIZER;
@end
```

**Comment by karwa** [comment]
@slavapestov @DougGregor

**Comment by karwa** [comment]
Pinging @DougGregor @slavapestov @jrose-apple 

Is there a reason why we shouldn't allow this?

**Comment by jrose-apple** [comment]
I'm pretty sure this does need to go through the evolution process, but I'm not a core team member. It seems to me that "protocols can't go inside other types" is a simpler rule than "protocols can't go inside generic types". But you're not the only person who's asked for it.

PrintAsObjC of nested types [is a mess](https://bugs.swift.org/browse/SR-10043) and I don't know what we're going to do about it.

**Comment by jrose-apple** [comment]
Some test ideas: a protocol inside another protocol (should fail, I think), a protocol in local scope, and a protocol inside a struct in local scope.

**Comment by karwa** [comment]
@jrose-apple It's the same restriction we had on _all_ nested types prior to Swift 3.1.

The decl/nesting test includes the protocol-inside-protocol. Nothing is allowed to be nested inside a protocol (including other protocols, but also structs/enums/classes). It also has a protocol-in-local-scope test, and I can add one for protocol-in-struct-in-local-scope.

We could relax the restriction on things being nested _inside of protocols_ if the nested types do not capture associated types from their parent, but that just creates yet more complex rules and is probably worth separate consideration. Allowing captures would need significant design work and probably some form of generalised existentials. Consider:

```swift
protocol WorkerProtocol {
	associatedtype Result
    
	enum Status {
 		case notStarted
		case working
		case finished(Result?)
	}
	var status: Status { get }
}

class IntWorker: WorkerProtocol { typealias Result = Int }
class StringWorker: WorkerProtocol { typealias Result = String }
```

We would need to invent the concept of a `(WorkerProtocol where Result == Int).Status`, and maybe more advanced cases like `(WorkerProtocol where Result: Numeric).Status`. We could transform `Status` in to a generic type (`Status<Int>/Status<Numeric>`), but we'd need generalised existentials and protocol self-conformance for that (it's basically like `Optional<Numeric>` or `Array<Numeric>`).

**Comment by jrose-apple** [comment]
> We could perhaps do some kind of compromise where protocols can be nested inside generic types (and capture them), but cannot themselves be parameterised.

FWIW this is equivalent to protocols with generic parameters (just as expressive and just as hard to implement), so I don't think this is an interesting point in the design space.

**Comment by slavapestov** [comment]
This post outlines a reasonable interpretation for nominal types nested in protocols: https://forums.swift.org/t/namespacing-protocols-to-other-types/7328/28

Generalized existentials are not required (but would complement the feature since it's pretty much orthogonal).

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by CodaFi** [comment]
@karwa Do you have the time to get this through Swift Evolution or should I close this down and convert it to an SR so somebody else can take a look?

**Comment by karwa** [comment]
@codafi yes, I’d be willing to take this through swift-evo. I’ll write a proposal either today or tomorrow.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**